RSDK-2389 Support reconfiguration of modules

benjirewis opened the pull request:
RSDK-2389

Main changes:
- Adds `Reconfigure` to module manager
- Adds `Remove` to module manager
- Refactors module manager to handle reconfiguration
- Adds `diffModules` to make `DiffConfigs` sensitive to changes in `modules` array
- Calls modular `Add`, `Reconfigure` and `Remove` from local robot `Reconfigure` to allow dynamic updates to the config's `modules` array
- Handles resources orphaned by removed/reconfigured modules in `Reconfigure`

Secondary changes:
- Adds test for module manager `Reconfigure`
- Adds test for module manager `Remove`
- Adds test for modular orphaned resources
- Adds test for mixed orphaned resources
- Modifies `TestDiffConfigs` to test module changes
- Adds `nolint:dupl` to some duplicated code

benjirewis commented:
I still need to test this feature more (particularly with C++ and Python modules), so I'll keep this as a draft for now. Any early feedback is welcome, though.

benjirewis commented on module/modmanager/manager.go:
```diff
@@ -335,6 +418,19 @@ func (m *module) startProcess(ctx context.Context, parentAddr string, logger gol
+		return err
+	}
+
+	utils.UncheckedErrorFunc(func() error { return os.Remove(m.addr) })
```
It might be better to remove the module's `.sock` file here instead of in `module.go` or the analogous `module.py` and `module/service.cpp`; Python and C++ were both missing this logic. I could add it to both SDKs, but it might make more sense to remove the `.sock` file where we're managing modules. 

With this change, reconfiguration now works correctly in Python (updated my larger testing comment).

Otterverse replied:
As noted on the change in module.go, I think removal of the socket should remain the responsibility of the process that created it in the first place. Having a backup check here isn't a bad idea though, but... as it causes logging if/when it was already cleaned up properly, I'd suggest expanding the lambda here to do a quick check. E.g. only remove if the file still exists, so there's one less spurious "error" showing in the logs.

benjirewis replied:
That sounds good to me. Expanded the closure here to check if the file exists before removing.

Otterverse commented on module/module.go:
```diff
@@ -177,7 +177,6 @@ func (m *Module) Start(ctx context.Context) error {
 	m.activeBackgroundWorkers.Add(1)
 	utils.PanicCapturingGo(func() {
 		defer m.activeBackgroundWorkers.Done()
-		defer utils.UncheckedErrorFunc(func() error { return os.Remove(m.addr) })
```
I think this should remain here. The module process creates the socket file, and so should remove it. But it's okay to have a fallback function attempt to cleanup after it in the parent too.

Otterverse replied:
Actually, as this sometimes logs errors when the file is already missing, it may be worth expanding this slightly to only attempt the removal if the "correct" stop sequence didn't remove it already.

benjirewis replied:
Expanded to only attempt removal if the .sock file still exists.

Otterverse reviewed (changes requested):
I've got two minor things to address (comments inline) but otherwise this looks great! Well done!

benjirewis commented on module/modmanager/manager.go:
```diff
@@ -115,12 +125,82 @@ func (mgr *Manager) Add(ctx context.Context, cfg config.Module) error {
+		return errors.Errorf("cannot remove module %s as it does not exist", modName)
+	}
+
+	if mod.process != nil {
```
Good catch. Removed.

cheukt commented on module/modmanager/manager.go:
```diff
@@ -115,12 +125,80 @@ func (mgr *Manager) Add(ctx context.Context, cfg config.Module) error {
+		return errors.WithMessage(err, "error while stopping module "+modName)
+	}
+
+	// Do not close connection or deregister resources if module is being reconfigured.
```
why not deregister? there's no guarantee that the updated module will serve the same models

benjirewis replied:
good point; I don't think we need special logic for resource registration in reconfiguration.

cheukt replied:
what would happen if we tried to reconfigure a module that no longer serves the model that the previous module served? I assume we try to build it once but then it'd get orphaned.

I wonder if we actually need to readd the resources at a higher level - basically flag that a rebuild is needed as part of the reconfiguration process but with the awareness that it shouldn't be a reconfigure request

cheukt reviewed (commented):
generally looks good - tho do have a question around de-registering/re-registering model handlers

benjirewis commented on module/modmanager/manager.go:
```diff
@@ -335,6 +417,28 @@ func (m *module) startProcess(ctx context.Context, parentAddr string, logger gol
+	// Stop managed process and swallow 143 errors. Some stopped modules may return
+	// "exit status 143" indicating a graceful exit after SIGTERM.
+	if err := m.process.Stop(); err != nil &&
+		!strings.Contains(err.Error(), gracefullyExitedMessage) {
```
This [Python SDK PR](https://github.com/viamrobotics/viam-python-sdk/pull/250) modified how Python modules stop to be slightly more graceful. They return `exit status 143`, which is indicative of a graceful shutdown after reception of a `SIGTERM`. We could add this error swallowing to `goutils`, too.

Otterverse replied:
So it's a long, long established standard that 0 is "exited normally" in linux/unix process space. Why are we using a strange custom code? This feels very wrong to me, and nothing in the PR or the linked jira issue mentions anything about defining custom exit codes or behavior.

Otterverse replied:
Can you let me know how to reproduce this? I pulled down the latest python sdk, went into the module in the examples directory, ran `poetry install` and then `./run.sh /tmp/test.sock` and it started up using poetry, no issues. I then sent it a SIGTERM with `kill -TERM $PID` (with the actual pid.) It exited gracefully, and the exit code (determined with `echo $?`) was simply 0.

If I forced a kill with -KILL then I got code 137 (which is 128+ the signal (kill==9) that it died on.) That typically means a signal that is NOT handled correctly. 143 (128+15) would, I believe mean something similar, that things died because of a SIGTERM (15) signal, but the shutdown was NOT graceful/error-free.

Otterverse commented on module/module.go:
```diff
@@ -177,7 +177,14 @@ func (m *Module) Start(ctx context.Context) error {
 		defer m.activeBackgroundWorkers.Done()
-		defer utils.UncheckedErrorFunc(func() error { return os.Remove(m.addr) })
+		defer utils.UncheckedErrorFunc(func() error {
+			// Attempt to remove module's .sock file if parent did not remove it
```
[nit] Actually this should be triggering BEFORE the parent's removal attempt I believe. Code is fine, comment is just slightly wrong I think.

Otterverse reviewed (commented):
Pretty much good, minus one nit and a big question around why/how we are using special exit codes. I need to understand the reasoning there, because to me it seems like a really bad idea to do something so non-standard.

benjirewis reviewed (commented):
@cheukt and I discussed offline and thought it would be good to also handle resources orphaned by removed/reconfigured modules in this PR ([per this comment](https://github.com/viamrobotics/rdk/pull/2114#discussion_r1152446640)).

`modmanager.Reconfigure` and `modmanager.Remove` now return the resources they've orphaned, and we manually add those resources to `diff.Removed` in `Reconfigure`. These changes mean Go and Python modules can be removed and re-added at a later time and pick up their old resources (I further updated my testing comment [here](https://github.com/viamrobotics/rdk/pull/2114#issuecomment-1485830833)). 

Let me know if that design seems bad, @Otterverse; definitely open to alternative ideas. FYI, we recently [made a similar change](https://github.com/viamrobotics/rdk/pull/2137) for resources orphaned by their dependencies.

benjirewis commented on module/modmanager/manager.go:
```diff
@@ -121,6 +135,83 @@ func (mgr *Manager) Add(ctx context.Context, cfg config.Module) error {
+	}
+
+	if mod.conn != nil {
+		if err := mod.conn.Close(); err != nil {
```
Previously, I was not closing the gRPC connection to the module when reconfiguring. I think we actually have to close this connection, as when we `dial` again in `add`, the old `conn` is leaked.

Otterverse replied:
So if that IS actually happening, this probably isn't the right (or at least sufficient) fix. The actual resources from a module use the same connection as the module itself, so by closing it, it seems like you're now disconnecting all those resources, and above in the reconfigure, you're running an "AddResource()" loop, but ignoring the newly returned clients (using the new module's new connection) so this seems like it would orphan a lot of resource clients (e.g. resource manager nodes) with broken connections.

Otterverse replied:
Either those existing resources (that may already be in use by client code, or as dependencies of other resources) will reconnect to the new version of the module, or this is going to get a LOT more complicated, and new clients with new connections (as returned by mgr.AddResource()) will have to somehow be swapped out for the previous client instances.

Otterverse commented on module/modmanager/manager_test.go:
```diff
@@ -163,6 +163,31 @@ func TestModManagerFunctions(t *testing.T) {
+	// counter1 should still be provided by reconfigured module.
+	ok = mgr.IsModularResource(rNameCounter1)
+	test.That(t, ok, test.ShouldBeTrue)
+
```
Please add tests that the counter is still actually usable and working after reconfiguration, not just that it exists in the list. I'm definitely worried about this now that it's dealing more with orphans AND trying to close the old connection.

Otterverse commented on module/modmanager/manager.go:
```diff
@@ -121,6 +135,83 @@ func (mgr *Manager) Add(ctx context.Context, cfg config.Module) error {
+	}
+	handledResources := mod.resources
+
+	if err := mgr.remove(mod); err != nil {
+		return nil, err
+	}
+
+	if err := mgr.add(ctx, cfg); err != nil {
+		return nil, err
+	}
```
These shouldn't be returning nil... if either of these fail, the orphaned resources are still actually orphaned and need to be closed and cleaned up.

benjirewis replied:
Updated here and in a few other places to return all the module's handled resources.

Otterverse commented on module/modmanager/manager.go:
```diff
@@ -121,6 +135,83 @@ func (mgr *Manager) Add(ctx context.Context, cfg config.Module) error {
 	return nil
 }
 
+// Reconfigure reconfigures an existing resource module and returns the names of
+// now orphaned resources.
```
Good idea to actually handle orphans! But I think there's a lot of nuances to work through on this.

Otterverse reviewed (changes requested):
So, good catch/idea to actually deal more properly with orphans, but I think it may have opened up some more edge cases because of it. This is definitely a lot of complexity here, so it's entirely possible I'm misunderstanding or overlooking details you actually saw and considered, so please let me know if that's the case. Until then, I've highlighted my specific concerns inline.

cheukt commented:
a good test for reconfiguration would be (can probably be simplified): module base depends on rdk motor, some rdk thing that depends on the base. Reconfigure the module (still serves module base) and see if the rdk thing that depends on the base still works. 2nd case would be to reconfigure the module to one that doesn't serve the module base and observe behavior. And then reconfigure it back to something that does

benjirewis commented:
@Otterverse, as discussed offline, I no longer close the underlying module connection in `remove` and instead re-use the old connection in `add` when doing a `Reconfigure`. Also removed the swallowing of 143 errors; will continue investigation into `poetry` issues as part of RSDK-2425. 

@cheukt, added a more complex test of orphaned resources where modular resources depend on RDK resources and vice-versa.

Otterverse commented on module/modmanager/manager_test.go:
```diff
@@ -67,9 +67,15 @@ func TestModManagerFunctions(t *testing.T) {
+	oldConn := mod.conn
+	err = mod.dial(mod.conn)
+	test.That(t, err, test.ShouldBeNil)
+	test.That(t, mod.conn, test.ShouldResemble, oldConn)
```
ShouldResemble doesn't check identity, so it could be a different connection with the same properties and still pass. I think directly using test.ShouldEqual is actually what you want.

benjirewis replied:
Good point; fixed.

Otterverse reviewed (approved):
LGTM, save for one slight test modification that's needed first.

benjirewis commented on module/modmanager/manager.go:
```diff
@@ -62,22 +70,19 @@ type Manager struct {
-		if mod.process != nil {
-			err = multierr.Combine(err, mod.process.Stop())
-		}
+		err = multierr.Combine(err, mgr.remove(mod, false))
```
`Close` should use `remove`, too. 

I don't think this `Close` will leave us with "orphaned" or un`Close`d resources. This method is only called [here](https://github.com/viamrobotics/rdk/blob/main/robot/impl/local_robot.go#L188). "Orphaned" resources really only apply when reconfiguring a robot, not closing one. And, these resources will have already been `Close`d when `Close` is called on the resource manager [a few lines before](https://github.com/viamrobotics/rdk/blob/main/robot/impl/local_robot.go#L185).

cheukt commented on robot/impl/local_robot_test.go:
```diff
@@ -2389,3 +2392,264 @@ func TestOrphanedResources(t *testing.T) {
+	m, err := r.ResourceByName(motor.Named("m"))
+	test.That(t, err, test.ShouldBeNil)
+
+	// Assert that motor 'm' can still be used.
```
I think we actually want to test `m1` here. I'd like to see a more substantive test - and you may need to create a custom implementation of something that depends on the `g` so that we can see that the child resource is behaving correctly (a generic resource that calls `g` and returns the result is good enough)

benjirewis replied:
Sounds good. Replaced `m1` with a `doodad` that depends on `g` and has a method `doThroughGizmo` that can make gRPC calls through `g`; everything seems to work.

cheukt reviewed (changes requested):
looks really good! just an ask for bit more substantive test

cheukt reviewed (approved):
LGTM
